src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+  UnionTypeNode
+} from '../types'
+import { TypeError } from './internalTypeErrors'
+import {
+  formatTypeString,
+  lookupType,
+  lookupTypeAlias,
+  pushEnv,
+  RETURN_TYPE_IDENTIFIER,
+  setDeclKind,
+  setType,
+  setTypeAlias,
+  source1TypeOverrides,
+  tAny,
+  tBool,
+  tFunc,
+  tNumber,
+  tPrimitive,
+  tString,
+  tUndef,
+  tUnion,
+  tVoid,
+  typeAnnotationKeywordToBasicTypeMap
+} from './utils'
+
+/**
+ * Entry function for type error checker.
+ * Checks programs for type errors, and returns the program with all TS-related nodes removed.
+ */
+export function checkForTypeErrors(
+  program: NodeWithTypeAnnotation<es.Program>,
+  context: Context
+): es.Program {
+  // Deep copy type environment to avoid modifying type environment in the context,
+  // which might affect the type inference checker
+  const env: TypeEnvironment = cloneDeep(context.typeEnvironment)
+  // Override predeclared function types
+  for (const [name, type] of source1TypeOverrides) {
+    setType(name, type, env)
+  }
+  try {
+    typeCheckAndReturnType(program, context, env)
+    return removeTSNodes(program)
+  } catch (error) {
+    // Catch-all for thrown errors
+    // (either errors that cause early termination or errors that should not be reached logically)
+    console.error(error)
+    context.errors.push(
+      error instanceof TypeError
+        ? error
+        : new TypeError(
+            program,
+            'Uncaught error during typechecking, report this to the administrators!\n' +
+              error.message
+          )
+    )
+    return removeTSNodes(program)
+  }
+}
+
+/**
+ * Recurses through the given node to check for any type errors,
+ * then returns the node's inferred/declared type.
+ * Any errors found are added to the context.
+ */
+function typeCheckAndReturnType(
+  node: NodeWithTypeAnnotation<es.Node>,
+  context: Context,
+  env: TypeEnvironment
+): Type {
+  switch (node.type) {
+    case 'Literal': {
+      // Infers type
+      if (node.value === undefined) {
+        return tUndef
+      }
+      if (node.value === null) {
+        // TODO: Handle null literal types for Source 2 and above
+        return tAny
+      }
+      const typeOfLiteral = typeof node.value as PrimitiveType
+      if (Object.values(PrimitiveType).includes(typeOfLiteral)) {
+        return tPrimitive(typeOfLiteral)
+      }
+      throw new TypeError(node, 'Unknown literal type')
+    }
+    case 'Identifier': {
+      const varName = node.name
+      const varType = lookupType(varName, env)
+      if (varType) {
+        return varType as Type
+      } else {
+        context.errors.push(new UndefinedVariable(varName, node))
+        return tAny
+      }
+    }
+    case 'Program':
+    case 'BlockStatement': {
+      let returnType: Type = tVoid
+      pushEnv(env)
+
+      if (node.type === 'Program') {
+        // Import statements should only exist in program body
+        handleImportDeclarations(node, context, env)
+      }
```

Comment:
If `import` statements that are not in the top-level program body are already flagged as an error by the parser, maybe we don't need to specifically perform this check? My concern here is that this is an implicit dependency on the parser. What if one day the parser is changed to allow `import` statements in functions? The `import` statements would silently fail. I think we can trust the parser with the responsibility of determining whether `import` statements are valid.

Reply:
Turns out removing this check actually causes type errors.